packages/node_modules/@node-red/nodes/core/network/21-httprequest.js

```diff
@@ -311,79 +336,96 @@ module.exports = function(RED) {
-            if (proxyConfig && proxyConfig.credentials && opts.proxy == proxyConfig.url) {
-                var proxyUsername = proxyConfig.credentials.username || '';
-                var proxyPassword = proxyConfig.credentials.password || '';
-                if (proxyUsername || proxyPassword) {
-                    opts.headers['proxy-authorization'] =
-                        'Basic ' +
-                        Buffer.from(proxyUsername + ':' + proxyPassword).toString('base64');
                 }
             }
             if (tlsNode) {
-                tlsNode.addTLSOptions(opts);
+                opts.https = {};
+                tlsNode.addTLSOptions(opts.https);
             } else {
                 if (msg.hasOwnProperty('rejectUnauthorized')) {
-                    opts.rejectUnauthorized = msg.rejectUnauthorized;
+                    opts.https = { rejectUnauthorized: msg.rejectUnauthorized };
                 }
             }
-            request(opts, function(err, res, body) {
-                if(err){
-                    if(err.code === 'ETIMEDOUT' || err.code === 'ESOCKETTIMEDOUT') {
-                        node.error(RED._("common.notification.errors.no-response"), msg);
-                        node.status({fill:"red", shape:"ring", text:"common.notification.errors.no-response"});
-                    }else{
-                        node.error(err,msg);
-                        node.status({fill:"red", shape:"ring", text:err.code});
-                    }
-                    msg.payload = err.toString() + " : " + url;
-                    msg.statusCode = err.code;
-                    nodeSend(msg);
-                    nodeDone();
-                }else{
-                    msg.statusCode = res.statusCode;
-                    msg.headers = res.headers;
-                    msg.responseUrl = res.request.uri.href;
-                    msg.payload = body;
-                    msg.redirectList = redirectList;
-
-                    if (msg.headers.hasOwnProperty('set-cookie')) {
-                        msg.responseCookies = extractCookies(msg.headers['set-cookie']);
-                    }
-                    msg.headers['x-node-red-request-node'] = hashSum(msg.headers);
-                    // msg.url = url;   // revert when warning above finally removed
-                    if (node.metric()) {
-                        // Calculate request time
-                        var diff = process.hrtime(preRequestTimestamp);
-                        var ms = diff[0] * 1e3 + diff[1] * 1e-6;
-                        var metricRequestDurationMillis = ms.toFixed(3);
-                        node.metric("duration.millis", msg, metricRequestDurationMillis);
-                        if (res.client && res.client.bytesRead) {
-                            node.metric("size.bytes", msg, res.client.bytesRead);
-                        }
+            got(url,opts).then(res => {
+                msg.statusCode = res.statusCode;
```

Comment:
As I think that @k-toumura said before, we would like to need the starting and ending time of the http request when we investigate logs of Node-RED flow to call REST APIs. We found that the `got` module is suitable to fill the requirement. Can we add the code like the following to output trace log?

```
log.trace('http request: nodeid=' + this.id + ' _msgid=' + msg._msgid
        + ' url=' + res.url + ' start=' + res.timings.start + ' end=' + res.timings.end);
```

Reply:
- We can also extend metric logs, such as:
```javascript
node.metric("duration.millis", msg, res.timings.phases.total);
node.metric("duration.wait.millis", msg, res.timings.phases.wait);
//... and others (dns, tcp, tls, request, firstByte, download)
```
- If httprequest node emit a message that contains entire `timings` object, following nodes can use this timing values in application (or use them for debugging purpose).
```javascript
msg.timings = res.timings
```

Reply:
I'm aware the current code doesn't yet do the metrics reporting the old node did. That is on my list to fix and yes, got makes it much easier to do then the old code.

But I'm not sure about adding timings to the message - we don't do that with any other node. I'm also not keen in exposing things the module gives us directly - because if we ever have to move off the got module, then it's even more work to reproduce the same behaviour.

As it stands, we're still blocked on this PR until we can get proxy support working. The issue i had raised on hpagent has not been answered - so need to look at a different plan for it.

Reply:
@knolleary @k-toumura Thank you for your comments. After supporting proxy in the new http request node, we can test it using our corporate proxy.

Reply:
could this be optional behind the metrics option in settings ?

Reply:
It could, but that doesn't change the fact it's becomes part of the public API of the node and puts us on the hook to ensure its always there